### PR TITLE
Convert isExactType check to JIT/EE interface call

### LINKED ISSUE #97134: JIT does not recognize typeof(X) == o.GetType() checks as intrinsics with R2R
Repro:
```csharp
using System;
using System.Runtime.CompilerServices;

public sealed class Test96876
{
    public static void Main()
    {
        foo<Test96876>(new DateTime[1]);
    }

    // Validate that the type equality involving shared array types is handled correctly
    // in shared generic code.
    [MethodImpl(MethodImplOptions.NoInlining)]
    static bool foo<T>(DateTime[] list) => typeof(T[]) == list.GetType();
}
```

## Actual result:

`foo<__Canon>` is optimized to return constant for JIT, but does full type check for R2R.

## Expected result

`foo<__Canon>` is optimized to return constant for both JIT and R2R.

Optimized JIT code:
```
G_M10478_IG02:  ;; offset=0x0000
       cmp      byte  ptr [rdx], dl
       xor      eax, eax
                                                ;; size=4 bbWeight=1 PerfScore 3.25
G_M10478_IG03:  ;; offset=0x0004
       ret
```

Optimized R2R code:
```
G_M000_IG01:                ;; offset=0x0000
       push     rbx
       sub      rsp, 48
       mov      qword ptr [rsp+0x28], rcx
       mov      rbx, rdx

G_M000_IG02:                ;; offset=0x000D
       call     [CORINFO_HELP_READYTORUN_GENERIC_HANDLE]
       cmp      qword ptr [rbx], rax
       sete     al
       movzx    rax, al

G_M000_IG03:                ;; offset=0x001C
       add      rsp, 48
       pop      rbx
       ret
```

### PR DESCRIPTION
**Pull request by jkotas**
The approximation of isExactType from class flags had false positives (correctness) and false negatives issues. Converting it to JIT/EE interface method fixes them both.

Fixes #97134

### CODE REVIEW

**Comment by jkotas** [comment]
cc @dotnet/jit-contrib

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -13597,29 +13597,9 @@ methodPointerInfo* Compiler::impAllocateMethodPointerInfo(const CORINFO_RESOLVED
...
-        CORINFO_CLASS_HANDLE arrayElementHandle = nullptr;
-        CorInfoType          type               = info.compCompHnd->getChildType(classHnd, &arrayElementHandle);
-
-        if ((type == CORINFO_TYPE_CLASS) || (type == CORINFO_TYPE_VALUECLASS))
-        {
-            return impIsClassExact(arrayElementHandle);
-        }
-    }
-    return false;
+    return info.compCompHnd->isExactType(classHnd);
```
Comment (EgorBo) [inline src/coreclr/jit/importer.cpp:13602]: I presume we can remove impIsClassExact and inline this call everywhere but it's very nit

**Review by EgorBo — APPROVED** [review]
It's a coincidence, but in https://github.com/dotnet/runtime/pull/97387/files I currently have a failing test that fails because IsExactClass returns true for a class with [TypeIdentifier] and I assume this PR will fix it. The reason current jit's cast expansion logic is not hitting this issue becuase it doesn't try to expand `CORINFO_HELP_ISINSTANCEOFANY` for exact classes

**Comment by EgorBo** [comment]
Does `IsExactType` return true for `int[]`?

**Comment by jkotas** [comment]
> Does IsExactType return true for int[]?

It does not and it cannot not. `int[]`, `uint[]` and arrays of `int`/`uint` backed enums are interchangeable.

**Comment by jkotas** [comment]
> I assume no and wonder if it results in some jit-diffs 

The JIT approximation returned false for int[] as well. 

I agree that it is interesting to get jit-diff.

**Comment by EgorBo** [comment]
There is also `getExactClasses` API which is implemented only for NativeAOT at the moment - I wonder if it doesn't respect type variance too. JIT might use that API to optimize casts and it might skip the fallback call because it assumes it's the only type that can pass the cast.

**Comment by jkotas** [comment]
> JIT might use that API to optimize casts and it might skip the fallback call 

Do you mean this place https://github.com/dotnet/runtime/blob/d250dcc2deae28fa9726ecad78dddf614b1420a8/src/coreclr/jit/importer.cpp#L5483-L5484 . The code checks whether the implementation is actually exact. I do not see a problem there. (There may be a naming confusion - `getExactClasses` vs. `isExactType` do not use "exact" to mean the same thing.)

**Comment by EgorBo** [comment]
> > JIT might use that API to optimize casts and it might skip the fallback call
> 
> Do you mean this place
> 
> https://github.com/dotnet/runtime/blob/d250dcc2deae28fa9726ecad78dddf614b1420a8/src/coreclr/jit/importer.cpp#L5483-L5484
> 
> ? The code checks whether the implementation is actually exact. I do not see a problem there. (There may be a naming confusion - `getExactClasses` vs. `isExactType` do not use "exact" to mean the same thing.)

Yes, I meant that case - I agree it's ok since it checks isExact explicitly 👍 

> There may be a naming confusion 

Agree, `getExactClasses` was named like that as an opposite to `getLikelyClasses`

**Comment by EgorBo** [comment]
@MihuBot

**Comment by EgorBo** [comment]
@jkotas an unrelated question:

Currently, for
```cs
var result = (MySealedClass)obj;
```
jit emits something like this:
```cs
if (obj.GetType() == typeof(MySealedClass))
    result = obj;
else
   CORINFO_HELP_CHKCASTCLASS(obj, MySealedClass)
   ret3 // always throw anyway
```

Can we hit here the same issue we hit recently with version bubbles in R2R if `MySealedClass` becomes non-sealed in a new version? or `isExactClass` won't return true for it even if has `sealed`?

**Comment by jkotas** [comment]
> isExactClass won't return true for it even if has sealed?

Yes. IsEffectivelySealed for R2R returns true only if the type is in the same version bubble:

https://github.com/dotnet/runtime/blob/d250dcc2deae28fa9726ecad78dddf614b1420a8/src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/DevirtualizationManager.cs#L18-L21

**Comment by EgorBo** [comment]
No jit-diffs https://github.com/MihuBot/runtime-utils/issues/228#issue-2097236648

**Comment by jkotas** [comment]
> the build-in-progress jobs seem to stuck? 10h for the NativeAOT one

`Received request to deprovision: The request was cancelled by the remote provider.`

These legs passed in previous runs. I do not think we need to retry.

**Pull request merged**